### PR TITLE
AGENT-348: Warn when information in install-config is ignored

### PR DESCRIPTION
**Pull request by pawanpinjarkar**
Log warnings for the irrelevant config data provided in install-config.yaml for the agent-based installer.

To Test, run 
`/usr/local/go/bin/go test -timeout 30s -run ^TestInstallConfigLoad/valid_configuration_for_baremetal_platform_for_HA_cluster_-_deprecated_and_unused_fields$ github.com/openshift/installer/pkg/asset/agent -v`

The test will print the warnings and the config data ignored.

Signed-off-by: Pawan Pinjarkar <ppinjark@redhat.com>

### CODE REVIEW

File: pkg/asset/agent/installconfig.go [hunk 1]
```diff
@@ -120,3 +124,172 @@ func (a *OptionalInstallConfig) ClusterName() string {
...
+		logrus.Warnf(fmt.Sprintf("%s is ignored", fieldPath))
+	}
+
+	switch installConfig.Platform.Name() {
+	case baremetal.Name:
+		baremetal := installConfig.Platform.BareMetal
+		// +kubebuilder:default="qemu:///system". Set from generic install config code
+		if baremetal.LibvirtURI != "qemu:///system" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "LibvirtURI")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.LibvirtURI))
```
Comment (zaneb) [inline pkg/asset/agent/installconfig.go:182]: This is only used for the bootstrap VM, which people (hopefully) know we don't need. So we could probably safely reduce this to debug level.

File: pkg/asset/agent/installconfig.go [hunk 2]
```diff
@@ -120,3 +124,172 @@ func (a *OptionalInstallConfig) ClusterName() string {
...
+			fieldPath := field.NewPath("Platform", "Baremetal", "ClusterProvisioningIP")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.ClusterProvisioningIP))
+		}
+		if baremetal.DeprecatedProvisioningHostIP != "" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "ProvisioningHostIP")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.DeprecatedProvisioningHostIP))
+		}
+		if baremetal.BootstrapProvisioningIP != "" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "BootstrapProvisioningIP")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.BootstrapProvisioningIP))
```
Comment (zaneb) [inline pkg/asset/agent/installconfig.go:194]: Also bootstrap-only, so could be debug.

File: pkg/asset/agent/installconfig.go [hunk 3]
```diff
@@ -120,3 +124,172 @@ func (a *OptionalInstallConfig) ClusterName() string {
...
+			fieldPath := field.NewPath("Platform", "Baremetal", "BootstrapOSImage")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.BootstrapOSImage))
+		}
+		if baremetal.ClusterOSImage != "" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "ClusterOSImage")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.ClusterOSImage))
+		}
+		if baremetal.BootstrapExternalStaticIP != "" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "BootstrapExternalStaticIP")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.BootstrapExternalStaticIP))
```
Comment (zaneb) [inline pkg/asset/agent/installconfig.go:288]: Could be debug

File: pkg/asset/agent/installconfig.go [hunk 4]
```diff
@@ -120,3 +124,172 @@ func (a *OptionalInstallConfig) ClusterName() string {
...
+		}
+		if baremetal.BootstrapExternalStaticGateway != "" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "BootstrapExternalStaticGateway")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.BootstrapExternalStaticGateway))
```
Comment (zaneb) [inline pkg/asset/agent/installconfig.go:292]: Could be debug

File: pkg/asset/agent/installconfig.go [hunk 5]
```diff
@@ -120,3 +124,172 @@ func (a *OptionalInstallConfig) ClusterName() string {
...
+			}
+		}
+
+		if baremetal.DefaultMachinePlatform != nil {
+			fieldPath := field.NewPath("Platform", "Baremetal", "DefaultMachinePlatform")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.DefaultMachinePlatform))
+		}
+		if baremetal.BootstrapOSImage != "" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "BootstrapOSImage")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.BootstrapOSImage))
```
Comment (zaneb) [inline pkg/asset/agent/installconfig.go:280]: Could be debug

File: pkg/asset/agent/installconfig.go [hunk 6]
```diff
@@ -120,3 +124,172 @@ func (a *OptionalInstallConfig) ClusterName() string {
...
+		}
+		if baremetal.ClusterOSImage != "" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "ClusterOSImage")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.ClusterOSImage))
```
Comment (zaneb) [inline pkg/asset/agent/installconfig.go:284]: This is ignored even in IPI now, so we probably don't need to check it at all.

File: pkg/asset/agent/installconfig.go [hunk 7]
```diff
@@ -120,3 +124,172 @@ func (a *OptionalInstallConfig) ClusterName() string {
...
+		}
+		if baremetal.BootstrapExternalStaticIP != "" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "BootstrapExternalStaticIP")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.BootstrapExternalStaticIP))
+		}
+		if baremetal.BootstrapExternalStaticGateway != "" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "BootstrapExternalStaticGateway")
+			logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.BootstrapExternalStaticGateway))
+		}
+	}
```
Comment (zaneb) [inline pkg/asset/agent/installconfig.go:294]: We also ignore the following fields in Platform.VSphere:

* VCenter
* Username
* Password
* Datacenter
* DefaultDatastore
* Folder
* Cluster
* ResourcePool
* ClusterOSImage
* DefaultMachinePlatform
* Network
* DIskType
* VCenters
* FailureDomains

**Review by zaneb — COMMENTED** [review]
Other fields we ignore:

* Publish
* BootstrapInPlace.InstallationDisk
* Capabilities
* FeatureSet

File: pkg/asset/agent/installconfig.go (hunk 7 above)
Comment (pawanpinjarkar) [reply pkg/asset/agent/installconfig.go:362]: Done

File: pkg/asset/agent/installconfig.go [hunk 8]
```diff
@@ -120,3 +125,239 @@ func (a *OptionalInstallConfig) ClusterName() string {
...
+	// "External" is the default set from generic install config code
+	if installConfig.Publish != "External" {
+		fieldPath := field.NewPath("Publish")
+		logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, installConfig.Publish))
+	}
+	if installConfig.CredentialsMode != "" {
+		fieldPath := field.NewPath("CredentialsMode")
+		logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, installConfig.CredentialsMode))
+	}
+	if installConfig.BootstrapInPlace.InstallationDisk != "" {
```
Comment (zaneb) [inline pkg/asset/agent/installconfig.go:355]:
```suggestion
	if installConfig.BootstrapInPlace != nil && installConfig.BootstrapInPlace.InstallationDisk != "" {
```
This is failing unit tests due to a nil pointer deref.

File: pkg/asset/agent/installconfig.go [hunk 9]
```diff
@@ -120,3 +125,239 @@ func (a *OptionalInstallConfig) ClusterName() string {
...
+		fieldPath := field.NewPath("BootstrapInPlace", "InstallationDisk")
+		logrus.Warnf(fmt.Sprintf("%s: %s is ignored", fieldPath, installConfig.BootstrapInPlace.InstallationDisk))
+	}
+	if installConfig.Capabilities != &(types.Capabilities{}) {
+		fieldPath := field.NewPath("Publish")
```
Comment (zaneb) [inline pkg/asset/agent/installconfig.go:360]: Capabilities

File: pkg/asset/agent/installconfig.go [hunk 10]
```diff
@@ -120,3 +125,239 @@ func (a *OptionalInstallConfig) ClusterName() string {
...
+		if baremetal.BootstrapExternalStaticGateway != "" {
+			fieldPath := field.NewPath("Platform", "Baremetal", "BootstrapExternalStaticGateway")
+			logrus.Debugf(fmt.Sprintf("%s: %s is ignored", fieldPath, baremetal.BootstrapExternalStaticGateway))
+		}
+	case vsphere.Name:
+		vspherePlatform := installConfig.Platform.VSphere
+
+		if vspherePlatform.VCenter != "" {
+			fieldPath := field.NewPath("Platform", "VSphere", "VCenter")
+			logrus.Debugf(fmt.Sprintf("%s: %s is ignored", fieldPath, vspherePlatform.VCenter))
```
Comment (zaneb) [inline pkg/asset/agent/installconfig.go:291]: vSphere ones should all be `Warnf` I think.

**Review by zaneb — COMMENTED** [review]
Couple of minor things, but this looks almost ready.

**Pull request merged**